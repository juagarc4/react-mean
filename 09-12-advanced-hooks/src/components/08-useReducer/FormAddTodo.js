import React from 'react'
import { useForm } from 'hooks/useForm'

export const FormAddTodo = ({ handleAddTodo }) => {
  const [{ description }, handleInputChange, reset] = useForm({
    description: '',
  })

  const handleSubmit = (e) => {
    e.preventDefault()
    if (description.trim().length <= 1) {
      return
    }
    const newTodo = {
      id: new Date().getTime(),
      description: description,
      done: false,
    }

    handleAddTodo({ newTodo })
    reset()
  }

  return (
    <>
      <h4>Add TODO</h4>
      <hr />
      <form onSubmit={handleSubmit}>
        <input
          type="text"
          name="description"
          placeholeder="Aprender..."
          autoComplete="off"
          className="form-control"
          onChange={handleInputChange}
          value={description}
        ></input>
        <button type="submit" className="btn btn-outline-primary mt-1">
          Add
        </button>
      </form>
    </>
  )
}
