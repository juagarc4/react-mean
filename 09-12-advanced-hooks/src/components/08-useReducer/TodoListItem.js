import React from 'react'

export const TodoListItem = ({ todo, index, handleDelete, handleToggle }) => {
  return (
    <li className="list-group-item">
      <p
        className={`${todo.done && 'complete'}`}
        onClick={() => handleToggle(todo.id)}
      >
        {index + 1}. {todo.description}
      </p>
      <button className="btn btn-danger" onClick={() => handleDelete(todo.id)}>
        Delete
      </button>
    </li>
  )
}
