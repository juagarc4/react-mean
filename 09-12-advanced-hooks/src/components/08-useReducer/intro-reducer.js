const initialState = [
  {
    id: 1,
    todo: 'Buy milk',
    done: false,
  },
]

const todoReducer = (state = initialState, action) => {
  if (action?.type === 'add') {
    return [...state, action.payload]
  }
  return state
}

let todos = todoReducer()

const newTodo = {
  id: 2,
  todo: 'Buy bread',
  done: false,
}

const addToDo = {
  type: 'add',
  payload: newTodo,
}

todos = todoReducer(todos, addToDo)

console.log(todos)
