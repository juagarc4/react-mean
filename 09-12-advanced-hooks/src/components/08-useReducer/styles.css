body {
  padding: 50px;
}
.list-group-item {
  display: flex;
  align-items: baseline;
  justify-content: space-between;
}
.list-group-item p {
  cursor: pointer;
}
.complete {
  text-decoration: line-through;
}
