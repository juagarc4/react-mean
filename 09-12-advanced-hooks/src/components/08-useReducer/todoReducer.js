export const todoReducer = (state = [], action) => {
  switch (action.type) {
    case 'add':
      return [...state, action.payload]
    case 'remove':
      return state.filter((todo) => todo.id !== action.payload)
    case 'toggle':
      return state.map((todo) => {
        return todo.id === action.payload
          ? { ...todo, donde: !todo.done }
          : todo
      })
    default:
      return state
  }
}
