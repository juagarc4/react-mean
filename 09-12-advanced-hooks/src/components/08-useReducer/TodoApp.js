import React, { useReducer, useEffect } from 'react'
import { todoReducer } from './todoReducer'
import './styles.css'
import { useForm } from 'hooks/useForm'

const init = () => {
  return JSON.parse(localStorage.getItem('todos')) || []
}
export const TodoApp = () => {
  const [todos, dispatch] = useReducer(todoReducer, [], init)

  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(todos))
  }, [todos])

  const [{ description }, handleInputChange, reset] = useForm({
    description: '',
  })

  const handleDelete = (todoId) => {
    const action = {
      type: 'remove',
      payload: todoId,
    }
    dispatch(action)
  }

  const handleToogle = (todoId) => {
    const action = {
      type: 'toggle',
      payload: todoId,
    }
    dispatch(action)
  }

  const handleSubmit = (e) => {
    e.preventDefault()
    if (description.trim().length <= 1) {
      return
    }
    const newTodo = {
      id: new Date().getTime(),
      description: description,
      done: false,
    }

    const action = {
      type: 'add',
      payload: newTodo,
    }

    dispatch(action)
    reset()
  }
  return (
    <>
      <h1>TodoApp ({todos.length}) </h1>
      <hr />
      <div className="row">
        <div className="col-7">
          <ul className="list-group list-group-flush">
            {todos.map((todo, i) => {
              return (
                <li key={todo.id} className="list-group-item">
                  <p
                    className={`${todo.done && 'complete'}`}
                    onClick={() => handleToogle(todo.id)}
                  >
                    {i + 1}. {todo.description}
                  </p>
                  <button
                    className="btn btn-danger"
                    onClick={() => handleDelete(todo.id)}
                  >
                    Delete
                  </button>
                </li>
              )
            })}
          </ul>
        </div>
        <div className="col-5">
          <h4>Add TODO</h4>
          <hr />
          <form onSubmit={handleSubmit}>
            <input
              type="text"
              name="description"
              placeholeder="Aprender..."
              autoComplete="off"
              className="form-control"
              onChange={handleInputChange}
              value={description}
            ></input>
            <button type="submit" className="btn btn-outline-primary mt-1">
              Add
            </button>
          </form>
        </div>
      </div>
    </>
  )
}
