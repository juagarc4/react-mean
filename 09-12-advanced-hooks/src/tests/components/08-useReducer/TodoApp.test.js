import React from 'react'
import { shallow, mount } from 'enzyme'
import { TodoApp } from 'components/08-useReducer/TodoApp'

import { testTodos } from 'tests/fixtures/testTodos'
import { act } from '@testing-library/react'

describe('Tests in component <TodoApp/>', () => {
  const wrapper = shallow(<TodoApp />)
  Storage.prototype.setItem = jest.fn(() => {})

  test('Should show the component properly', () => {
    expect(wrapper).toMatchSnapshot()
  })

  test('Should add a Todo', () => {
    const wrapper = mount(<TodoApp />)
    act(() => {
      wrapper.find('FormAddTodo').prop('handleAddTodo')(testTodos[0])
      wrapper.find('FormAddTodo').prop('handleAddTodo')(testTodos[1])
    })

    expect(wrapper.find('h1').text().trim()).toBe('TodoApp (2)')
    expect(localStorage.setItem).toHaveBeenCalledTimes(2)
  })

  test('Should delete a Todo', () => {
    wrapper.find('FormAddTodo').prop('handleAddTodo')(testTodos[0])
    wrapper.find('TodoList').prop('handleDelete')(testTodos[0].id)
    expect(wrapper.find('h1').text().trim()).toBe('TodoApp (0)')
  })
})
