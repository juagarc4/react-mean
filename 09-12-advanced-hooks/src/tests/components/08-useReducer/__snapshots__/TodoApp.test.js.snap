// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in component <TodoApp/> Should show the component properly 1`] = `
<Fragment>
  <h1>
    TodoApp (
    0
    ) 
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <FormAddTodo
        handleAddTodo={[Function]}
      />
    </div>
  </div>
</Fragment>
`;
